File: express/src/service/assistance-service.js
import db from "../application/database.js";
import { ResponseError } from "../response/response-error.js";
import "dotenv/config";
import { logger } from "../application/logging.js";
import { createAssistanceToolsValidation } from "../validation/assistance-tools-validation.js";
import { validate } from "../validation/validation.js";

const getAssistanceById = async (id) => {
  try {
    const [rows] = await db.promise().query(
      `SELECT 
          assistance.id, assistance.nama, assistance.koordinator, 
          assistance.sumber_anggaran, assistance.total_anggaran, 
          assistance.tahun_pemberian, 
          assistance_tools.kuantitas, 
          tools.nama_item, tools.harga, tools.deskripsi
        FROM assistance
        LEFT JOIN assistance_tools ON assistance.id = assistance_tools.assistance_id
        LEFT JOIN tools ON assistance_tools.tools_id = tools.id
        WHERE assistance.id = ?`,
      [id]
    );

    if (rows.length === 0) {
      throw new ResponseError(404, "Assistance not found");
    }

    const result = {
      id: rows[0].id,
      nama: rows[0].nama,
      koordinator: rows[0].koordinator,
      sumber_anggaran: rows[0].sumber_anggaran,
      total_anggaran: rows[0].total_anggaran,
      tahun_pemberian: rows[0].tahun_pemberian,
      tools: [],
    };

    rows.forEach((row) => {
      if (row.nama_item) {
        result.tools.push({
          nama_item: row.nama_item,
          kuantitas: row.kuantitas,
          harga: row.harga,
          deskripsi: row.deskripsi,
        });
      }
    });

    return result;
  } catch (error) {
    logger.error(error);
    throw new ResponseError(500, error.message);
  }
};

const createAssistanceTools = async (req, res) => {
  try {
    const data = validate(createAssistanceToolsValidation, req);
    const result = await db
      .promise()
      .query("INSERT INTO assistance_tools SET ?", [data]);
    return result;
  } catch (error) {
    throw new ResponseError(400, error.message);
  }
};

export default { getAssistanceById, createAssistanceTools };

File: flask/src/service/assistance_service.py
from response.response_error import ResponseError
from validation.validation import validate
from validation.assistance_validation import CreateAssistanceToolsValidation
from application.database import get_cursor, db_connection

def get_assistance_tools(id):
    cursor = get_cursor()
    if cursor is None:
        raise ResponseError(500, "Error obtaining database connection")
    
    try:
        query = """
            SELECT 
                assistance.id, assistance.nama, assistance.koordinator, 
                assistance.sumber_anggaran, assistance.total_anggaran, 
                assistance.tahun_pemberian, 
                assistance_tools.kuantitas, 
                tools.nama_item, tools.harga, tools.deskripsi
            FROM assistance
            LEFT JOIN assistance_tools ON assistance.id = assistance_tools.assistance_id
            LEFT JOIN tools ON assistance_tools.tools_id = tools.id
            WHERE assistance.id = %s
        """

        cursor.execute(query, (id,))
        rows = cursor.fetchall()

        if not rows:
            raise ResponseError(404, "Assistance not found")
        
        assistance = rows[0]
        payload = {
            'id': assistance['id'],
            'nama': assistance['nama'],
            'koordinator': assistance['koordinator'],
            'sumber_anggaran': assistance['sumber_anggaran'],
            'total_anggaran': assistance['total_anggaran'],
            'tahun_pemberian': assistance['tahun_pemberian'],
            'tools': []
        }

        for row in rows:
            if row['nama_item']:
                payload['tools'].append({
                    'kuantitas': row['kuantitas'],
                    'nama_item': row['nama_item'],
                    'harga': row['harga'],
                    'deskripsi': row['deskripsi']
                })

        return payload
    except ResponseError as e:
        raise e
    except Exception as e:
        raise ResponseError(500, str(e))
    finally:
        if cursor:
            cursor.close()
    
def create_assistance_tools(req):
    cursor = get_cursor()
    if cursor is None:
        raise ResponseError(500, "Error obtaining database connection")
    
    try:
        data = validate(CreateAssistanceToolsValidation, req)

        assistance_id = data.assistance_id
        tools_id = data.tools_id
        kuantitas = data.kuantitas

        query = """
            INSERT INTO assistance_tools (assistance_id, tools_id, kuantitas)
            VALUES (%s, %s, %s)
        """

        cursor.execute(query, (assistance_id, tools_id, kuantitas))
        db_connection.commit()

        return "oke"
    except ResponseError as e:
        raise e
    except Exception as e:
        raise ResponseError(500, str(e))
    finally:
        if cursor:
            cursor.close()